### PR TITLE
cephadm: add support for sidecar containers

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Depends on #54536

Add the initial support for creating sidecar containers. Sidecar containers are tightly coupled with a primary service container. This PR only adds the ability to create and delete sidecars via systemd service unit files. As part of the work updating systemd integration to use additional services the init containers are moved into a dedicated systemd service. The `cephadm ls` output and the mgr are ignorant of sidecars and init containers.

The iscsi service had a special care where the tcmu runner container was started as part of the unit.run file in the background. This has been converted to a sidecar container. The lack of insight into the sidecars should be no worse than before as the tcmu runner container was largely "invsible" to systemd & the mgr before this change.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s>
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Updated to disable `Restart=on-failure` line in init container service template.

File: src/cephadm/cephadm.py [hunk 1]
```diff
@@ -4152,12 +4153,22 @@ def command_rm_daemon(ctx):
...
+    terminate_service(ctx, unit_name)
+
+    # clean up any extra systemd unit files
+    sd_path_info = systemd_unit.sidecars_from_dropin(
+        systemd_unit.PathInfo(ctx.unit_dir, ident), missing_ok=True
+    )
+    for sc_unit in sd_path_info.sidecar_unit_files.values():
+        terminate_service(ctx, sc_unit.name)
+        unlink_file(sc_unit, missing_ok=True)
+    terminate_service(ctx, sd_path_info.init_ctr_unit_file.name)
```
Comment (adk3798) [inline src/cephadm/cephadm.py:4165]: does this imply there is only ever one service for the init containers(s) while the sidecars can have multiple?

File: src/cephadm/cephadmlib/runscripts.py [hunk 2]
```diff
@@ -90,6 +88,23 @@ def write_service_scripts(
                 meta['ports'] = []
         metaf.write(json.dumps(meta, indent=4) + '\n')
 
+        # init-container commands
+        if init_containers:
+            initf = estack.enter_context(write_new(initctr_file_path))
+            _write_init_containers_script(ctx, initf, init_containers)
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/cephadm/cephadmlib/runscripts.py:94]: meant to post this earlier. Had already been fixed iirc.

**Review by adk3798 — APPROVED** [review]

File: src/cephadm/cephadmlib/daemon_identity.py [hunk 3]
```diff
@@ -48,9 +57,35 @@ def container_name(self) -> str:
...
+        template_terms = [framework, self.fsid, category]
+        instance_terms = [self.daemon_type]
+        instance_terms.append(
+            f'{self.daemon_id}:{suffix}' if suffix else self.daemon_id
+        )
+        instance_terms.append(extension)
+        # use a comprehension to filter out terms that are blank
+        base = '-'.join(v for v in template_terms if v)
+        svc = '.'.join(v for v in instance_terms if v)
+        return f'{base}@{svc}'
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/cephadm/cephadmlib/daemon_identity.py:80]: Will do.

File: src/cephadm/cephadm.py (hunk 1 above)
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:4165]: That's correct. Init containers are grouped together under a single service. Each sidecar has a unique service. It's not set in stone but that's the design at the moment.

File: src/cephadm/cephadmlib/daemon_identity.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/cephadm/cephadmlib/daemon_identity.py:80]: Done

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-01-02_14:31:29-orch:cephadm-wip-adk-testing-2024-01-02-0614-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2024-01-03_10:25:36-orch:cephadm-wip-adk-testing-2024-01-02-0614-distro-default-smithi/

After reruns: 3 failed/dead jobs:

- 1 instance of https://tracker.ceph.com/issues/62958
- 1 failure in test_iscsi_container timing out on rhel 8.6. Stil haven't figured this out totaly but might have something to do with the fact it seems to be trying to set the OS as bot hrhel 8.6 and centos 8.stream (title for that job is orch:cephadm/workunits/{0-distro/rhel_8.6_container_tools_rhel8 agent/on mon_election/connectivity task/test_iscsi_container/{centos_8.stream_container_tools test_iscsi_container}}) still needs more investigation, but the failure predates this run, so unrelated to PRs in the run.
- 1 failure in test_orch_cli task. After investigating, it turned out the agent was getting marked down as it failed to send messages to the mgr because it was configured with the incorrect port (was trying to send to port 7150, but mgr was listening on 7152). Will also need a further look, but I don't think it's caused by any PRs in this run either

Overall, nothing to block merging.

**Pull request merged**